File: test/Models/pheno_pkg/src/simplace/Phyllochron.java
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import javafx.util.*;
import java.time.LocalDateTime;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom.Element;


public class Phyllochron extends FWSimComponent
{
    private FWSimVariable<Double> lincr;
    private FWSimVariable<Double> ldecr;
    private FWSimVariable<Double> pdecr;
    private FWSimVariable<Double> pincr;
    private FWSimVariable<Double> kl;
    private FWSimVariable<Double> pTQhf;
    private FWSimVariable<Double> B;
    private FWSimVariable<Double> p;
    private FWSimVariable<String> choosePhyllUse;
    private FWSimVariable<Double> areaSL;
    private FWSimVariable<Double> areaSS;
    private FWSimVariable<Double> lARmin;
    private FWSimVariable<Double> lARmax;
    private FWSimVariable<Double> sowingDensity;
    private FWSimVariable<Double> lNeff;
    private FWSimVariable<Double> fixPhyll;
    private FWSimVariable<Double> leafNumber;
    private FWSimVariable<Double> ptq;
    private FWSimVariable<Double> gAImean;
    private FWSimVariable<Double> phyllochron;

    public Phyllochron(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public Phyllochron(){
        super();
    }
    @Override
    protected void process()
    {
        double tfixPhyll = fixPhyll.getValue();
        double tleafNumber = leafNumber.getValue();
        double tlincr = lincr.getValue();
        double tldecr = ldecr.getValue();
        double tpdecr = pdecr.getValue();
        double tpincr = pincr.getValue();
        double tptq = ptq.getValue();
        double tgAImean = gAImean.getValue();
        double tkl = kl.getValue();
        double tpTQhf = pTQhf.getValue();
        double tB = B.getValue();
        double tp = p.getValue();
        String tchoosePhyllUse = choosePhyllUse.getValue();
        double tareaSL = areaSL.getValue();
        double tareaSS = areaSS.getValue();
        double tlARmin = lARmin.getValue();
        double tlARmax = lARmax.getValue();
        double tsowingDensity = sowingDensity.getValue();
        double tlNeff = lNeff.getValue();
        double tphyllochron;
        double gaiLim;
        double LAR;
        tphyllochron = 0.0d;
        LAR = 0.0d;
        gaiLim = tlNeff * ((tareaSL + tareaSS) / 10000.0d) * tsowingDensity;
        if (tchoosePhyllUse == "Default")
        {
            if (tleafNumber < tldecr)
            {
                tphyllochron = tfixPhyll * tpdecr;
            }
            else if ( tleafNumber >= tldecr && tleafNumber < tlincr)
            {
                tphyllochron = tfixPhyll;
            }
            else
            {
                tphyllochron = tfixPhyll * tpincr;
            }
        }
        if (tchoosePhyllUse == "PTQ")
        {
            if (tgAImean > gaiLim)
            {
                LAR = (tlARmin + ((tlARmax - tlARmin) * tptq / (tpTQhf + tptq))) / (tB * tgAImean);
            }
            else
            {
                LAR = (tlARmin + ((tlARmax - tlARmin) * tptq / (tpTQhf + tptq))) / (tB * gaiLim);
            }
            tphyllochron = 1.0d / LAR;
        }
        if (tchoosePhyllUse == "Test")
        {
            if (tleafNumber < tldecr)
            {
                tphyllochron = tp * tpdecr;
            }
            else if ( tleafNumber >= tldecr && tleafNumber < tlincr)
            {
                tphyllochron = tp;
            }
            else
            {
                tphyllochron = tp * tpincr;
            }
        }
        phyllochron.setValue(tphyllochron, this);
    }

    @Override
    protected void init()
    {
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new Phyllochron(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("clincr", "Leaf number above which the phyllochron is increased by Pincr", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 30.0, 8.0, this));
        addVariable(FWSimVariable.createSimVariable("cldecr", "Leaf number up to which the phyllochron is decreased by Pdecr", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 100.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("cpdecr", "Factor decreasing the phyllochron for leaf number less than Ldecr", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 10.0, 0.4, this));
        addVariable(FWSimVariable.createSimVariable("cpincr", "Factor increasing the phyllochron for leaf number higher than Lincr", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 10.0, 1.5, this));
        addVariable(FWSimVariable.createSimVariable("ckl", "Exctinction Coefficient", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 50.0, 0.45, this));
        addVariable(FWSimVariable.createSimVariable("cpTQhf", "Slope to intercept ratio for Phyllochron  parametrization with PhotoThermal Quotient", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("cB", "Phyllochron at PTQ equal 1", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 20.0, this));
        addVariable(FWSimVariable.createSimVariable("cp", "Phyllochron (Varietal parameter)", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 120.0, this));
        addVariable(FWSimVariable.createSimVariable("cchoosePhyllUse", "Switch to choose the type of phyllochron calculation to be used", DATA_TYPE.CHAR, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, "Default", this));
        addVariable(FWSimVariable.createSimVariable("careaSL", " Area Leaf", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("careaSS", "Area Sheath", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("clARmin", "LAR minimum", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("clARmax", "LAR maximum", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("csowingDensity", "Sowing Density", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("clNeff", "Leaf Number efficace", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("ifixPhyll", "Sowing date corrected Phyllochron", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 10000.0, 5.0, this));
        addVariable(FWSimVariable.createSimVariable("sleafNumber", "Actual number of phytomers", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 25.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("sptq", "Photothermal quotient ", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 10000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("sgAImean", "Green Area Index", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 10000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("sphyllochron", " the rate of leaf appearance ", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 1000, null, this));

        return iFieldMap;
    }
}

File: test/Models/pheno_pkg/src/cs/Phyllochron.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class Phyllochron
{
    private double _lincr;
    public double lincr
    {
        get { return this._lincr; }
        set { this._lincr= value; } 
    }
    private double _ldecr;
    public double ldecr
    {
        get { return this._ldecr; }
        set { this._ldecr= value; } 
    }
    private double _pdecr;
    public double pdecr
    {
        get { return this._pdecr; }
        set { this._pdecr= value; } 
    }
    private double _pincr;
    public double pincr
    {
        get { return this._pincr; }
        set { this._pincr= value; } 
    }
    private double _kl;
    public double kl
    {
        get { return this._kl; }
        set { this._kl= value; } 
    }
    private double _pTQhf;
    public double pTQhf
    {
        get { return this._pTQhf; }
        set { this._pTQhf= value; } 
    }
    private double _B;
    public double B
    {
        get { return this._B; }
        set { this._B= value; } 
    }
    private double _p;
    public double p
    {
        get { return this._p; }
        set { this._p= value; } 
    }
    private string _choosePhyllUse;
    public string choosePhyllUse
    {
        get { return this._choosePhyllUse; }
        set { this._choosePhyllUse= value; } 
    }
    private double _areaSL;
    public double areaSL
    {
        get { return this._areaSL; }
        set { this._areaSL= value; } 
    }
    private double _areaSS;
    public double areaSS
    {
        get { return this._areaSS; }
        set { this._areaSS= value; } 
    }
    private double _lARmin;
    public double lARmin
    {
        get { return this._lARmin; }
        set { this._lARmin= value; } 
    }
    private double _lARmax;
    public double lARmax
    {
        get { return this._lARmax; }
        set { this._lARmax= value; } 
    }
    private double _sowingDensity;
    public double sowingDensity
    {
        get { return this._sowingDensity; }
        set { this._sowingDensity= value; } 
    }
    private double _lNeff;
    public double lNeff
    {
        get { return this._lNeff; }
        set { this._lNeff= value; } 
    }
    public Phyllochron() { }
    
    public void  Calculate_phyllochron(PhenologyState s, PhenologyState s1, PhenologyRate r, PhenologyAuxiliary a)
    {
        //- Name: Phyllochron -Version: 1.0, -Time step: 1
        //- Description:
    //            * Title: Phyllochron Model
    //            * Author: Pierre Martre
    //            * Reference: Modeling development phase in the 
    //                Wheat Simulation Model SiriusQuality.
    //                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    //            * Institution: INRA Montpellier
    //            * Abstract: Calculate different types of phyllochron 
        //- inputs:
    //            * name: fixPhyll
    //                          ** description : Sowing date corrected Phyllochron
    //                          ** inputtype : variable
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** default : 5.0
    //                          ** min : 0.0
    //                          ** max : 10000.0
    //                          ** unit : °C d leaf-1
    //                          ** uri : some url
    //            * name: leafNumber
    //                          ** description : Actual number of phytomers
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 25.0
    //                          ** unit : leaf
    //                          ** uri : some url
    //            * name: lincr
    //                          ** description : Leaf number above which the phyllochron is increased by Pincr
    //                          ** inputtype : parameter
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** default : 8.0
    //                          ** min : 0.0
    //                          ** max : 30.0
    //                          ** unit : leaf
    //                          ** uri : some url
    //            * name: ldecr
    //                          ** description : Leaf number up to which the phyllochron is decreased by Pdecr
    //                          ** inputtype : parameter
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 100.0
    //                          ** unit : leaf
    //                          ** uri : some url
    //            * name: pdecr
    //                          ** description : Factor decreasing the phyllochron for leaf number less than Ldecr
    //                          ** inputtype : parameter
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** default : 0.4
    //                          ** min : 0.0
    //                          ** max : 10.0
    //                          ** unit : -
    //                          ** uri : some url
    //            * name: pincr
    //                          ** description : Factor increasing the phyllochron for leaf number higher than Lincr
    //                          ** inputtype : parameter
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** default : 1.5
    //                          ** min : 0.0
    //                          ** max : 10.0
    //                          ** unit : -
    //                          ** uri : some url
    //            * name: ptq
    //                          ** description : Photothermal quotient 
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 10000.0
    //                          ** unit : MJ °C-1 d-1 m-2)
    //                          ** uri : some url
    //            * name: gAImean
    //                          ** description : Green Area Index
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 10000.0
    //                          ** unit : m2 m-2
    //                          ** uri : some url
    //            * name: kl
    //                          ** description : Exctinction Coefficient
    //                          ** inputtype : parameter
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** default : 0.45
    //                          ** min : 0.0
    //                          ** max : 50.0
    //                          ** unit : -
    //                          ** uri : some url
    //            * name: pTQhf
    //                          ** description : Slope to intercept ratio for Phyllochron  parametrization with PhotoThermal Quotient
    //                          ** inputtype : parameter
    //                          ** parametercategory : genotypic
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : °C d leaf-1
    //                          ** uri : some url
    //            * name: B
    //                          ** description : Phyllochron at PTQ equal 1
    //                          ** inputtype : parameter
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** default : 20.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : °C d leaf-1
    //                          ** uri : some url
    //            * name: p
    //                          ** description : Phyllochron (Varietal parameter)
    //                          ** inputtype : parameter
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** default : 120.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : °C d leaf-1
    //                          ** uri : some url
    //            * name: choosePhyllUse
    //                          ** description : Switch to choose the type of phyllochron calculation to be used
    //                          ** inputtype : parameter
    //                          ** parametercategory : species
    //                          ** datatype : STRING
    //                          ** default : Default
    //                          ** min : 
    //                          ** max : 
    //                          ** unit : -
    //                          ** uri : some url
    //            * name: areaSL
    //                          ** description :  Area Leaf
    //                          ** inputtype : parameter
    //                          ** parametercategory : genotypic
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : cm2
    //                          ** uri : some url
    //            * name: areaSS
    //                          ** description : Area Sheath
    //                          ** inputtype : parameter
    //                          ** parametercategory : genotypic
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : cm2
    //                          ** uri : some url
    //            * name: lARmin
    //                          ** description : LAR minimum
    //                          ** inputtype : parameter
    //                          ** parametercategory : genotypic
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : leaf-1 °C
    //                          ** uri : some url
    //            * name: lARmax
    //                          ** description : LAR maximum
    //                          ** inputtype : parameter
    //                          ** parametercategory : genotypic
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : leaf-1 °C
    //                          ** uri : some url
    //            * name: sowingDensity
    //                          ** description : Sowing Density
    //                          ** inputtype : parameter
    //                          ** parametercategory : genotypic
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : plant m-2
    //                          ** uri : some url
    //            * name: lNeff
    //                          ** description : Leaf Number efficace
    //                          ** inputtype : parameter
    //                          ** parametercategory : genotypic
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 1000.0
    //                          ** unit : leaf
    //                          ** uri : some url
        //- outputs:
    //            * name: phyllochron
    //                          ** description :  the rate of leaf appearance 
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 1000
    //                          ** unit :  °C d leaf-1
    //                          ** uri : some url
        double fixPhyll = a.fixPhyll;
        double leafNumber = s.leafNumber;
        double ptq = s.ptq;
        double gAImean = s.gAImean;
        double phyllochron;
        double gaiLim;
        double LAR;
        phyllochron = 0.0d;
        LAR = 0.0d;
        gaiLim = lNeff * ((areaSL + areaSS) / 10000.0d) * sowingDensity;
        if (choosePhyllUse == "Default")
        {
            if (leafNumber < ldecr)
            {
                phyllochron = fixPhyll * pdecr;
            }
            else if ( leafNumber >= ldecr && leafNumber < lincr)
            {
                phyllochron = fixPhyll;
            }
            else
            {
                phyllochron = fixPhyll * pincr;
            }
        }
        if (choosePhyllUse == "PTQ")
        {
            if (gAImean > gaiLim)
            {
                LAR = (lARmin + ((lARmax - lARmin) * ptq / (pTQhf + ptq))) / (B * gAImean);
            }
            else
            {
                LAR = (lARmin + ((lARmax - lARmin) * ptq / (pTQhf + ptq))) / (B * gaiLim);
            }
            phyllochron = 1.0d / LAR;
        }
        if (choosePhyllUse == "Test")
        {
            if (leafNumber < ldecr)
            {
                phyllochron = p * pdecr;
            }
            else if ( leafNumber >= ldecr && leafNumber < lincr)
            {
                phyllochron = p;
            }
            else
            {
                phyllochron = p * pincr;
            }
        }
        s.phyllochron= phyllochron;
    }
}